Fix reuse of client port on FreeBSD

ararslan opened the pull request:
Fixes #22302 

I think DragonFly BSD might do the same thing as Linux here but I can't confirm because I've yet to get Julia to build on DragonFly. This at least gets the distributed tests passing on FreeBSD.

amitmurthy commented:
Thinking a bit more, I realize we should replace the OS checks by checking if `SO_REUSEPORT ` is defined. Can have a C function that checks for it and use that both here and in the test script.

ararslan committed:
Add jl_has_so_reuseport to check whether SO_REUSEPORT is defined

Without this, the best proxy is to condition on the OS, which is less
reliable and may not properly cover all cases.

ararslan commented:
@amitmurthy Were you thinking something like this?

ararslan commented on base/distributed/managers.jl:
```diff
@@ -455,10 +455,10 @@ end
 
 function socket_reuse_port()
-    @static if is_linux() || is_apple()
+    if Bool(ccall(:jl_has_so_reuseport, Int32, ()))
```
Realized that this can be `ccall(:jl_has_so_reuseport, Bool, ())`. I'll make that change if this approach is acceptable.

vtjnash replied:
`Bool` is a `Int8` (really `Int1`) type, so those wouldn't be equivalent. Although, I thought we deprecated this Bool constructor some years ago and required `!= 0` now?

ararslan replied:
I guess if this PR is good then I'll make it `!= 0`.

amitmurthy replied:
Also https://github.com/JuliaLang/julia/pull/22303/files#diff-5b6252dbf4f6ba35c7ee92dde3cb6451L836 can return `-1` now as Julia code is expected to check with `jl_has_so_reuseport`  before calling `jl_tcp_reuseport`

ararslan replied:
Right, I was talking more specifically about the return type of `jl_has_so_reuseport`, which will only ever be 0 or 1, so I figured that one was safe to convert to `Bool`.

amitmurthy commented on base/distributed/managers.jl:
```diff
@@ -455,10 +455,10 @@ end
+        # Linux requires the port to be bound before setting REUSEPORT, BSD after.
         is_linux() && bind_client_port(s)
         rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Void},), s.handle)
         if rc > 0  # SO_REUSEPORT is unsupported, just return the ephemerally bound socket
```
This check can now be removed.

ararslan replied:
Do you mean the entire conditional on `rc` or just the `rc > 0` section?

amitmurthy replied:
`rc > 0` section.

amitmurthy commented on base/distributed/managers.jl:
```diff
@@ -470,7 +470,7 @@ function socket_reuse_port()
             return TCPSocket()
         end
-        is_apple() && bind_client_port(s)
+        is_bsd() && bind_client_port(s)
```
Should change this to `!is_linux() && bind_client_port(s)`. 

Or ideally, few lines above to `bind_port = is_linux() ? true : false ; bind_port && bind_client_port(s)` and this line to `!bind_port && bind_client_port(s)`.

amitmurthy replied:
Just to ensure that if in the future,  Windows starts supporting reuseport, and requires early binding, we then have to change only the first line.

amitmurthy commented on test/distributed_exec.jl:
```diff
@@ -48,7 +48,7 @@ end
 # Test that the client port is reused. SO_REUSEPORT may not be supported on
 # all UNIX platforms, Linux kernels prior to 3.9 and older versions of OSX
-if is_unix()
+if ccall(:jl_has_so_reuseport, Int32, ()) == 1
```
The lines below were checking if reuseport is supported. Can be simplified now to directly call `reuseport_tests()` if supported else print an informational string.

amitmurthy commented on src/jl_uv.c:
```diff
@@ -823,6 +823,15 @@ JL_DLLEXPORT int jl_tcp_quickack(uv_tcp_t *handle, int on)
+
 JL_DLLEXPORT int jl_tcp_reuseport(uv_tcp_t *handle)
 {
 #if defined(SO_REUSEPORT)
```
Can return `-1` in the `#else` condition now.

ararslan replied:
I'm confused—this function could always return -1.

amitmurthy replied:
I meant
```
#if defined(SO_REUSEPORT)
......
#else
   return -1;
#endif
```

ararslan replied:
```diff
 JL_DLLEXPORT int jl_tcp_reuseport(uv_tcp_t *handle)
 {
 #if defined(SO_REUSEPORT)
     int fd = (handle)->io_watcher.fd;
     int yes = 1;
     if (setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, &yes, sizeof(yes))) {
         return -1;
     }
     return 0;
 #else
-    return 1;
+    return -1;
 #endif
 }
```
?

amitmurthy replied:
Yes.

ararslan replied:
Okay, will do. Thanks!

amitmurthy commented:
LGTM.

ararslan commented:
Thanks for all your help and guidance, @amitmurthy!
